src/test/java/fr/socrates/domain/candidate/CandidateServiceImplTest.java

```diff
@@ -46,5 +53,33 @@ public void should_find_a_candidate_by_email() throws Exception {
         assertThat(candidateService.findCandidateByEmail(email).isPresent()).isTrue();
     }
 
+    @Test
+    public void should_accept_to_add_candidate_if_he_doesnt_exist_before() throws Exception {
+        final String email = "john@doe.fr";
+        final boolean candidateAddingStatus = candidateService.add(Candidate.singleRoomWithEmail(email));
+        assertThat(candidateAddingStatus).isTrue();
+    }
+
+
+    @Test(expected = CandidateExistingException.class)
+    public void should_reject_candidate_when_already_exists() throws Exception {
+        final String email = "john@doe.fr";
+        candidateService.add(Candidate.singleRoomWithEmail(email));
+        candidateService.add(Candidate.singleRoomWithEmail(email));
+
+    }
+
+    @Test(expected = CandidatePersisteDataException.class)
+    public void should_manage_candidate_failure_in_persistence() throws Exception {
+        final String email = "john@doe.fr";
+        final Candidate candidate = Candidate.singleRoomWithEmail(email);
+        candidateService = new CandidateServiceImpl(candidateRepository);
+        when(candidateRepository.save(any())).thenReturn(false);
+        when(candidateRepository.findByCandidateID(any())).thenReturn(Optional.empty());
```

Comment:
That mock is really useful for that test?

Reply:
yes to test the exception CandidatePersisteDataException I have to mock save and findByCandidateID 
PS : we have to use findByEmail instead of findByCandidateID